### PR TITLE
[WIP] add kafka rule type to iot topic rules

### LINKED ISSUE #17506: aws_iot_topic_rule support for MSK
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

* Terraform v0.13+
+ provider.aws v3.27.0

AWS has launched the support for delivering message to MSK in IoT Actions, see
https://aws.amazon.com/about-aws/whats-new/2020/12/aws-iot-core-adds-the-ability-to-deliver-data-to-apache-kafka-clusters/ for more details.

Any plan to support this?

### PR DESCRIPTION
**Pull request by toddradigan**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #17506

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSIoTTopicRule_kafka'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSIoTTopicRule_kafka -timeout 180m
=== RUN   TestAccAWSIoTTopicRule_kafka
=== PAUSE TestAccAWSIoTTopicRule_kafka
=== CONT  TestAccAWSIoTTopicRule_kafka
    resource_aws_iot_topic_rule_test.go:287: Step 1/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_iot_topic_rule.rule will be updated in-place
          ~ resource "aws_iot_topic_rule" "rule" {
                id          = "test_rule_16e88"
                name        = "test_rule_16e88"
                tags        = {}
                # (5 unchanged attributes hidden)
        
              - kafka {
                  - acks                  = 0 -> null
                  - bootstrap_servers     = "b-1.localhost:9094" -> null
                  - compression_type      = "none" -> null
                  - destination_arn       = "[ redacted ]" -> null
                  - key_serializer        = "org.apache.kafka.common.serialization.StringSerializer" -> null
                  - security_protocol     = "SSL" -> null
                  - ssl_keystore          = "${get_secret('[ redacted ]', 'SecretBinary', '', '[ redacted ]')}" -> null
                  - ssl_keystore_password = "password" -> null
                  - topic                 = "fake_topic" -> null
                  - value_serializer      = "org.apache.kafka.common.serialization.ByteBufferSerializer" -> null
                }
              + kafka {
                  + acks                  = 1
                  + bootstrap_servers     = "b-1.localhost:9094"
                  + compression_type      = "none"
                  + destination_arn       = "[ redacted ]"
                  + key_serializer        = "org.apache.kafka.common.serialization.StringSerializer"
                  + security_protocol     = "SSL"
                  + ssl_keystore          = "${get_secret('[ redacted ], 'SecretBinary', '', '[ redacted ]')}"
                  + ssl_keystore_password = "password"
                  + topic                 = "fake_topic"
                  + value_serializer      = "org.apache.kafka.common.serialization.ByteBufferSerializer"
                }
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccAWSIoTTopicRule_kafka (24.82s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	26.579s
FAIL
```

### CODE REVIEW

**Comment by toddradigan** [comment]
What is the correct way to use aws `get_secrets(...)` in the `ssl_keystore` (and other) fields? ([dos ref](https://docs.aws.amazon.com/iot/latest/developerguide/apache-kafka-rule-action.html))

File: aws/resource_aws_iot_topic_rule_test.go [hunk 1]
```diff
@@ -861,6 +886,31 @@ resource "aws_iot_topic_rule" "rule" {
...
+  name        = "test_rule_%[1]s"
+  description = "Example rule"
+  enabled     = true
+  sql         = "SELECT * FROM 'topic/test'"
+  sql_version = "2015-10-08"
+  kafka {
+    destination_arn    = aws_vpc.iot_vpc.arn
+		topic							 = "fake_topic"
+		bootstrap_servers = "b-1.localhost:9094"
+		ssl_keystore 					= "${get_secret("aws_secretsmanager_secret.keystore.arn", "SecretBinary", "", "aws_iam_role.iot_role.arn")}"
```
Comment (mhaley-miovision) [inline aws/resource_aws_iot_topic_rule_test.go:907]: @toddradigan Do you just need to escape this? Is that your problem? Just ping me later if you want to chat more.
```suggestion
		ssl_keystore 					= "$${get_secret("${aws_secretsmanager_secret.keystore.arn}", "SecretBinary", "", "${aws_iam_role.iot_role.arn}")}"
```

**Comment by toddradigan** [comment]
For the kafka iot rule creation, it expects an iot vpc destination arn.  Is this something that can currently be created in terraform, or otherwise is there a vpc destination arn available in the environment used to run the acceptance tests?

**Comment by ax-jwinn** [comment]
> Pull request #21306 has significantly refactored the AWS Provider codebase. As a result, most PRs opened prior to the refactor now have merge conflicts that must be resolved before proceeding.
> 
> Specifically, PR #21306 relocated the code for all AWS resources and data sources from a single `aws` directory to a large number of separate directories in `internal/service`, each corresponding to a particular AWS service. This separation of code has also allowed for us to simplify the names of underlying functions -- while still avoiding namespace collisions.
> 
> We recognize that many pull requests have been open for some time without yet being addressed by our maintainers. Therefore, we want to make it clear that resolving these conflicts _in no way affects the prioritization_ of a particular pull request. Once a pull request has been prioritized for review, the necessary changes will be made by a maintainer -- either directly or in collaboration with the pull request author.
> 
> For a more complete description of this refactor, including examples of how old filepaths and function names correspond to their new counterparts: please refer to issue #20000.
> 
> For a quick guide on how to amend your pull request to resolve the merge conflicts resulting from this refactor and bring it in line with our new code patterns: please refer to our [Service Package Refactor Pull Request Guide](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/service-package-pullrequest-guide.md).

Is there any traction on this?  Or is it completely stale?  I am actually looking to use this functionality relatively quickly, am happy to contribute if necessary to get this across the finish line.

**Comment by SelfhostedPro** [comment]
Hey, just wanted to check and see if there are any updates on this or anything else that needs to be done?

**Comment by ewbankkit** [comment]
Superseded by #24395.
@toddradigan Thanks for the contribution :tada: :clap:.

**Pull request closed without merging**